libvast/src/system/node.cpp

```diff
@@ -575,68 +579,76 @@ node(node_actor::stateful_pointer<node_state> self, std::string name,
-    // bugs.
-    auto schedule_teardown = [&](const std::string& type_or_label) {
-      if (is_singleton(type_or_label)) {
-        if (auto component = self->state.registry.remove(type_or_label)) {
-          VAST_VERBOSE("{} schedules {} for shutdown", self, type_or_label);
-          self->demonitor(component->actor);
-          scheduled_for_teardown.push_back(std::move(component->actor));
-        }
-      } else if (auto components
-                 = self->state.registry.find_by_type(type_or_label);
-                 !components.empty()) {
-        VAST_VERBOSE("{} schedules {} {}(s) for shutdown", self,
-                     components.size(), type_or_label);
-        for (auto& component : components) {
-          if (auto removed = self->state.registry.remove(component)) {
-            self->demonitor(removed->actor);
-            scheduled_for_teardown.push_back(std::move(removed->actor));
-          }
-        }
+    self->state.tearing_down = true;
+    // Ignore duplicate EXIT messages except for hard kills.
+    self->set_exit_handler([=](const caf::exit_msg& msg) {
+      if (msg.reason == caf::exit_reason::kill) {
+        VAST_WARN("{} received hard kill and terminates immediately", self);
+        self->quit(msg.reason);
+      } else {
+        VAST_DEBUG("{} ignores duplicate EXIT message from {}", self,
+                   msg.source);
       }
-    };
-    // Terminate the accountant first because it acts like a source and may
-    // hold buffered data.
-    schedule_teardown("accountant");
-    // Take out the filesystem, which we terminate at the very end.
-    auto filesystem = deregister_component(self, "filesystem");
-    // Tear down the pipeline from source to sink. Note that the order is
-    // important here; the source must be shut down before the importer.
-    auto pipeline = {"source",  "importer", "disk-monitor", "index",
-                     "archive", "exporter", "sink"};
-    for (const auto* component : pipeline)
-      schedule_teardown(component);
-    // Now schedule all remaining components for termination.
-    auto& registry = self->state.registry;
-    std::vector<std::string> remaining;
-    remaining.reserve(registry.components().size());
-    for (const auto& [label, _] : registry.components())
-      remaining.push_back(label);
-    for (const auto& label : remaining)
-      schedule_teardown(label);
-    // Finally, bring down the filesystem.
-    scheduled_for_teardown.push_back(std::move(*filesystem));
+    });
+    std::vector<caf::actor> teardown_aux;
+    for (const auto& [_, comp] : self->state.registry.components()) {
+      self->demonitor(comp.actor);
+      // Ignore remote actors.
+      if (comp.actor->node() != self->node())
+        continue;
+      // Core components are terminated last.
+      if (comp.type == "archive" || comp.type == "filesystem"
+          || comp.type == "importer" || comp.type == "index")
```

Comment:
Please create some form of `is_core_component(std::string_view)` function.

Also, should the accountant be considered a core component? What about the transformer? Please add a comment to #1517 if necessary.

Maybe "core component" is the wrong name and it should be called "component requires sequential shutdown"?